src/upgrade.py

```diff
@@ -262,7 +276,29 @@ def pre_upgrade_check(self) -> None:
         See https://chat.canonical.com/canonical/pl/cmf6uhm1rp8b7k8gkjkdsj4mya
         """
         logger.debug("Running pre-upgrade checks")
-        # TODO: implement checks
-        # e.g.
-        # if health != green:
-        #     raise PrecheckFailed("Cluster is not healthy")
+
+        try:
+            if (health := self._charm.health.get(local_app_only=False)) != HealthColors.GREEN:
+                PrecheckFailed(
+                    f"Cluster not healthy: expected 'green', but '{health}' found instead"
+                )
+
+            online_nodes = ClusterTopology.nodes(
+                self._charm.opensearch,
+                True,
+                hosts=self._charm.alt_hosts,
+            )
+            if (
+                len([node for node in online_nodes if node.app_name == self._charm.app.name])
+                != self._charm.app.planned_units()
+            ):
+                PrecheckFailed("Not all units are online for this juju application")
+
+            if self.check_if_starting():
+                PrecheckFailed("Cluster is starting")
+
+            if not self._charm.backup.is_idle_or_not_set():
+                PrecheckFailed("Backup or restore is in progress")
+
+        except OpenSearchHttpError:
```

Comment:
We should ensure that every method making an http request gets retried 3 times or so, as to avoid failing for a temporary error

Reply:
@Mehdi-Bendriss:
* OpenSearchHealth.get: calls _health, with retries=3
* ClusterTopology.nodes: retries=3
* OpenSearch.is_idle_or_not_set: all API calls go to the same intermediary method that sets retries=6